File: frontend/src/model/questionnaires/questionnaireQuestion.ts
import { QuestionnaireQuestionOption, QuestionnaireQuestionOptionCreate } from './questionnaireQuestionOption';
import { QuestionnaireAnswer } from './questionnaireAnswer';

export enum QuestionnaireQuestionType {
  SINGLE_CHOICE = 0,
  FREE_TEXT = 1
}

export const QuestionnaireQuestionTypeNames: { [key in QuestionnaireQuestionType]: string } = {
  [QuestionnaireQuestionType.SINGLE_CHOICE]: 'Single choice',
  [QuestionnaireQuestionType.FREE_TEXT]: 'Freitext'
};

export interface QuestionnaireQuestion {
  id: number;
  questionnaire_id: string;
  order: number;
  question_text: string;
  is_mandatory: boolean;
  question_type: QuestionnaireQuestionType;
  answers?: QuestionnaireAnswer[];
  options?: QuestionnaireQuestionOption[];
}

export interface QuestionnaireQuestionCreate {
  question_text: string;
  is_mandatory: boolean;
  question_type: QuestionnaireQuestionType;
  options?: QuestionnaireQuestionOptionCreate[];
}

export interface QuestionnaireQuestionUpdate {
  id: number;
  questionnaire_id: string;
  order: number;
  question_text: string;
  is_mandatory: boolean;
  question_type: QuestionnaireQuestionType;
  answers?: QuestionnaireAnswer[];
  options?: QuestionnaireQuestionOption[];
}

File: learn-api/app/schemas/questionnaire_question.py
from enum import IntEnum
from pydantic import BaseModel
from typing import List, Optional

from app.schemas.questionnaire_answer import QuestionnaireAnswer
from app.schemas.questionnaire_question_option import (
    QuestionnaireQuestionOption,
    QuestionnaireQuestionOptionCreate,
)


class QuestionnaireQuestionType(IntEnum):
    SINGLE_CHOICE = 0
    FREE_TEXT = 1


class QuestionnaireQuestionBase(BaseModel):
    order: int
    question_text: str
    is_mandatory: bool = False
    question_type: QuestionnaireQuestionType
    answers: Optional[List[QuestionnaireAnswer]]
    options: Optional[List[QuestionnaireQuestionOption]]


class QuestionnaireQuestionCreate(QuestionnaireQuestionBase):
    questionnaire_id: Optional[int]
    options: Optional[List[QuestionnaireQuestionOptionCreate]]


class QuestionnaireQuestionUpdate(QuestionnaireQuestionBase):
    id: int
    questionnaire_id: Optional[int]
    order: Optional[int]
    question_text: Optional[str]
    question_type: Optional[QuestionnaireQuestionType]
    answers: Optional[List[QuestionnaireAnswer]]
    options: Optional[List[QuestionnaireQuestionOption]]


class QuestionnaireQuestionInDBBase(QuestionnaireQuestionBase):
    id: int
    questionnaire_id: int

    class Config:
        orm_mode = True


class QuestionnaireQuestionInDB(QuestionnaireQuestionInDBBase):
    pass


class QuestionnaireQuestion(QuestionnaireQuestionInDB):
    pass
